core/src/lib/channel.h

```diff
@@ -0,0 +1,385 @@
+/*
+   BAREOS® - Backup Archiving REcovery Open Sourced
+
+   Copyright (C) 2023-2023 Bareos GmbH & Co. KG
+
+   This program is Free Software; you can redistribute it and/or
+   modify it under the terms of version three of the GNU Affero General Public
+   License as published by the Free Software Foundation and included
+   in the file LICENSE.
+
+   This program is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+   Affero General Public License for more details.
+
+   You should have received a copy of the GNU Affero General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+   02110-1301, USA.
+*/
+
+#ifndef BAREOS_LIB_CHANNEL_H_
+#define BAREOS_LIB_CHANNEL_H_
+
+#include <condition_variable>
+#include <optional>
+#include <vector>
+#include <utility>
+
+namespace channel {
+// a simple single consumer/ single producer queue
+// Its composed of three parts: the input, the output and the
+// actual queue data.
+// Instead of directly interacting with the queue itself you instead
+// interact with either the input or the output.
+// This ensures that there is only one producer (who writes to the input)
+// and one consumer (who reads from the output).
+template <typename T> struct in;
+
+template <typename T> struct out;
+
+template <typename T> struct data {
+  // this contains the actual queue that the in and
+  // out channel use to communicate
+  friend struct in<T>;
+  friend struct out<T>;
+
+  data(std::size_t capacity);
+
+ private:
+  std::unique_lock<std::mutex> wait_for_readable();
+  std::unique_lock<std::mutex> wait_for_writable();
+
+  std::size_t size;
+  const std::size_t capacity;
+
+  // this should become a span once we switch to c++20
+  std::vector<T> storage;
```

Comment:
This is an out of date comment i believe.  In an earlier version the vector was used as fixed size ring buffer.  Since this is now not the case anymore, the comment should be deleted.

Reply:
already done in the refactor.